### DISCUSSION TITLE
How to deploy milvus with minimal memory for 6 billion data

### DISCUSSION THREAD

**Discussion by TianWuYuJiangHenShou** [milvus-io/milvus#30169] [Q&A and General discussion]
Existing 6 billion data (which may increase periodically in the future)
Collections: 
- one scalar field (varchar range from 20 to 50, maximum length not exceeding 4000, primary key),
- one vector field with dimension 512  (**immutable**)
- metric_type:JACCARD (**immutable**)
- index_type:BIN_IVF_FLAT
Generate a vector at once without deleting or modifying data, and perform frequent queries.
Deployment plan: k8s distributed deployment or data splitting using Docker Compose for multi node deployment

Question：
- The solution provided by Sizing Tool requires too much memory. Is there a simplified deployment plan？
- Current  Deployment plan:
  - K8s distributed deployment(our environment may not be able to use k8s,currently using slurm)
  - Split data into multiple nodes and deploy them separately using Docker Compose;then using flask for multi threaded Search

**Comment by yhmo** [comment]
The Sizing tool only calculates float vectors. Your dataset is binary vector, each dimension of a binary vector only occupies one bit. So, a 512dim binary vector only occupies 64 bytes.
For 6 billion 512dim binary vectors, the original size is 60 * 100M * 64 bytes = 375GB.

Milvus only provides two kinds of indexes for binary vector: BIN_FLAT and BIN_IVF_FLAT. Both of the two indexes' size are almost equal to the original data size.

For the varchar field, the required RAM is 60 * 100M * 50 bytes = 300GB.

So, your milvus server requires at least 700GB RAM to keep the index in memory for search. To avoid potential OOM, usually, we estimate the RAM capacity by using a safety ratio 1.5 ~ 2.0, which means we recommend allocating a totally 1TB~1.4TB RAM for your dataset.

In my opinion, a milvus cluster with 8 ~ 16 query nodes, each querynode 100 ~ 150GB RAM is ok.

**Reply by TianWuYuJiangHenShou** [reply]
@yhmo Thank you very much！
I have another question:
- Is there a memory limit for the milvus standalone container？

I deployed a milvus v2.3 standalone, the node has 504G RAM. However, when loading the collection, the milvus standalone container crashes when the RAM reaches 160G.

**Reply by xiaofan-luan** [reply]
What is the log message or error message when it crashed?
Did you get a physical machine with 504G memory?

For such large cluster, I would recommend to have 8 querynodes, each querynodes with 64GB memory

**Reply by TianWuYuJiangHenShou** [reply]
@xiaofan-luan yes,I try it with a physical machine with 504G memory.Before it crashed, the client kept retrying to connect the server, and the container crashed shortly.

**Reply by xiaofan-luan** [reply]
please offer server side log

**Reply by xiaofan-luan** [reply]
We can help you on investigating the root case
